specs/anchored-region/anchored-region.md

```diff
@@ -0,0 +1,260 @@
+
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalPositioningMode="dynamic"
+        verticalinset="true"
+    >
+      This overlaps the anchor and extends above or below it depending on available space.
+    </anchored-region>
+    ...stuff...
+</div>
+
+```
+
+A region renders above or below the anchor based on available space and is sized to match the available.
+
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalPositioningMode="dynamic"
+        verticalscalingenabled="true"
+    >
+       This region renders above or below the anchor based on available space and sizes itself to match.
+    </anchored-region>
+    ...stuff...
+</div>
+
+```
+
+A region that renders below the anchor until that space is less than 100px.
+```
+<div id="viewport">
+    ...stuff...
+    <button id="anchor">
+        Button is an anchor
+    </button>
+    <anchored-region
+        anchor="anchor"
+        verticalpositioningmode="dynamic"
+        verticaldefaultposition="below"
+        verticalthreshold="100"
+    >
+      This shows renders below the anchor as long as there is at least 100px available there.
+    </anchored-region>
+    ...stuff...
+</div>
+```
+
+The dimensions of the anchored region will match the dimensions of the content unless scaling is enabled on a particular axis (verticalscalingenabled & horizontalscalingenabled) in which case it will fill all available space between the anchor and viewport.
+
+The component allows users to set a "Positioning Mode" on each axis which defines how the component will behave:
+- 'uncontrolled':  The anchored region will appear as it normally would in document flow.
+- 'locktodefault': The anchored region will always be placed in the specified default position regardless of available space.  For example a menu that always opens upwards.
+- 'dynamic': The anchored region is placed relative to the anchor element based on how much space is available between it and the viewport.  For example a menu that opens opens upwards when it is near the bottom of the page, and downwards when near the top.
+- 'onetime': As the 'dynamic' option, but position is not changed after initial placement.
+
+### API
+
+*Component name:*
+- `fast-anchored-region`
+
+*Attributes:*
+- anchor - The html id of the HTMLElement used as the anchor around which the positioning region is placed.  This must be set for the component's positioning logic to be active.
+
+- viewport - Used to identify the HTMLElement used as the viewport the component uses to determine available layout space around the anchor element. Possible values are the id of the desired viewport element, '#document' or '#parent'.  The default value is '#parent'.
```

Comment:
Does this rely on the implementation including IDs in the DOM?

Reply:
Yes.  I'm open to other ideas.  The previous iteration used refs but that's a React thing.  I think we'd want to add the ability to pass references through JS as well, maybe not in first iteration tho. I'll add a mention of that.

Reply:
using IDs is a pretty idiomatic way to approach this I think - many native HTML elements rely on id associations.

Reply:
I'm fine with it, I just don't know if we should be very prescriptive about what those IDs should be or have defaults for them.

Reply:
Agreed - I noted above but "#parent" is a valid ID which *could* have unintentional effects. We probably want to avoid *any* default value here.